# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:

def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(432))

6
434


#### Saving to a Variable

In [2]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [6]:
def multiply(f,num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4) # lambda 4: 4 * 4

16

#### Returning a Lambda from a Function

In [7]:
def returnLambda(a, b):
    return lambda x,y: x + y + a + b

lambdafunction = returnLambda(10000, 1000)
print(lambdafunction(42, 20))

otheroption = returnLambda(-20, -10)
print(otheroption(20, 10))
print(otheroption(-100, -1000))












# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

11062
0
-1130


#### If Statements within Lambdas

In [8]:
posinegative = lambda num: 'positive' if num >= 0 else 'negative'

print(posinegative(-10))

print(posinegative(2))

print(posinegative(0))

def posineg(num):
    return 'positive' if num >= 0 else 'negative'



# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

negative
positive
positive
10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [13]:
f = lambda x: x**3

f(5)

125

In [130]:
animals = [
    {'family': 'Canidae', 'name': 'Fennec Fox', 'habitat': 'desert'}, 
    {'family': 'Felidae', 'name': 'Ocelot', 'given name': 'Babou', 'nickname': 'buyers remorse', 'habitat': 'jungle'},
    {'name': 'Orca'}, 
    {'name': 'Coyote'}
]
sorted(animals, key = lambda x: x['name'])


[{'name': 'Coyote'},
 {'family': 'Canidae', 'name': 'Fennec Fox', 'habitat': 'desert'},
 {'family': 'Felidae',
  'name': 'Ocelot',
  'given name': 'Babou',
  'nickname': 'buyers remorse',
  'habitat': 'jungle'},
 {'name': 'Orca'}]

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [18]:
def squared(num):
    if num < 10:
        return num**2
    else:
        return num
    
numbers = [12,83,2,65,5,3,9]

squared_numbers = list(map(squared, numbers))
print(squared_numbers)




#--------------------


# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num,num2
    
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)

[12, 83, 4, 65, 25, 9, 81]
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


In [19]:
def create_dict(name,breed):
    return {'name': name, 'breed': breed}

dog_names = ['Henry', 'Glacier', 'Wesley', 'Griffin', 'Nanuk']
dog_breeds = ['English Springer Spaniel', 'Samoyed', 'English Springer Spaniel', 'German Shepherd mix', 'Siberian Husky']

dog_dictionaries = list(map(create_dict, dog_names, dog_breeds))
print(dog_dictionaries)

[{'name': 'Henry', 'breed': 'English Springer Spaniel'}, {'name': 'Glacier', 'breed': 'Samoyed'}, {'name': 'Wesley', 'breed': 'English Springer Spaniel'}, {'name': 'Griffin', 'breed': 'German Shepherd mix'}, {'name': 'Nanuk', 'breed': 'Siberian Husky'}]


In [20]:
dogs = []
dog1 = create_dict('Henry', 'English Springer Spaniel')
dogs.append(dog1)
dogs

[{'name': 'Henry', 'breed': 'English Springer Spaniel'}]

#### Using Lambda's with Map

In [24]:
f = lambda x: x**3

nums = [12,83,2,65,5,3,9]
othernums = [2,4,6,3,5,1,0]

cubed = list(map(f, nums))
print(cubed)

nums = [12,83,2,65,5,3,9]
othernums = [2,4,6,3,5,1,0]

exponential = list(map(lambda x,y: x**y, nums, othernums))
print(exponential)


[1728, 571787, 8, 274625, 125, 27, 729]
[144, 47458321, 64, 274625, 3125, 3, 1]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [30]:
nums = [12,83,2,65,5,3,9]

x2sub1 = list(map(lambda x: x*2-1, nums))
x2sub1

[23, 165, 3, 129, 9, 5, 17]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [97]:
# filter(function, iterable)
# a filter object which you can easily convert to a list
# often your function will return True or False
# filter keeps the things in the iterable which returned True

animals = ['Fennec Fox', 'Red Fox', 'Arctic Fox', 'Orca', 'Bear', 'Ocelot']

def isItAFox(name):
    if 'f' in name.lower():
        return True
    else:
        return False
    
foxes = list(filter(isItAFox, animals))
foxes





#-------------------------







['Fennec Fox', 'Red Fox', 'Arctic Fox']

#### Using Lambda's with Filter()

In [ ]:
animals = ['Fennec Fox', 'Red Fox', 'Arctic Fox', 'Orca', 'Bear', 'Ocelot']

foxes = list(filter(lambda name: True if 'fox' in name.lower() else False, animals))
foxes



#-----------------



new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [38]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]
print(list(filter(lambda x: True if x < mean(nums) else False, nums)))


[2, 4.2, 1.6, 4.4]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [ ]:
# adding all numbers between 1 and n together

# basic way to write this function
f(n) = 1 + 2 + 3 + ... + n

# recursive way to write
f(n) = n + f(n-1)

# we're going to repeatedly call the same function from within itself
# until we hit a 'base case' which we define and tesll us when to stop repeatedly calling our function


#### Implementing a Base Case

In [41]:
def addNums(n):
    # base case
    if n <= 1:
        print('addNums(1) = 1')
        return n
    # recursive call
    else:
        print(f'addNums({n}) = {n} + addNums({n-1})')
        return n + addNums(n-1)

addNums(10)










addNums(10) = 10 + addNums(9)
addNums(9) = 9 + addNums(8)
addNums(8) = 8 + addNums(7)
addNums(7) = 7 + addNums(6)
addNums(6) = 6 + addNums(5)
addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


55

In [31]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = {num} + addNums({num - 1})")
        return num + addNums(num - 1)
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [46]:
# factorial: n! = n*n-1*n-2*...*2*1
# factorial recursively: n! = n * (n-1)

# our recursive call is going to be ^ that n * facto(n-1)
# our base case if our input n <= 1
def facto(n):
# base case - aka when do we stop
    if n <= 1:
        print(f'facto({n}) = 1')
        return 1
    # recursive call - aka when are we not done yet - when we have not opened our smallest doll
    else:
        print(f'facto{n} = {n} * facto{n-1})')
        return n * facto(n-1)
    
    
print(f'facto(5) final answer = {facto(5)}')

facto5 = 5 * facto4)
facto4 = 4 * facto3)
facto3 = 3 * facto2)
facto2 = 2 * facto1)
facto(1) = 1
facto(5) final answer = 120


In [2]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
factorial(5)


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [51]:
# f(n) = n-(n-1) - (n-1) - ... -2-1
# f(10) = 10-9-8-7-6-5-4-3-2-1

def subtracto(n):
    if n <= 1:
        print(f'sub({n}) = 1')
        return 1
    else:
        print(f'sub({n}) = {n} - sub({n-1})')
        return n - subtracto(n-1)
print (f'sub(10) final answer = {subtracto(10)})')

sub(10) = 10 - sub(9)
sub(9) = 9 - sub(8)
sub(8) = 8 - sub(7)
sub(7) = 7 - sub(6)
sub(6) = 6 - sub(5)
sub(5) = 5 - sub(4)
sub(4) = 4 - sub(3)
sub(3) = 3 - sub(2)
sub(2) = 2 - sub(1)
sub(1) = 1
sub(10) final answer = 5)


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [52]:
# the generator that we have all seen and use all the time: range()
print(list(range(1,10,2)))
for x in range(1,10,2):
    print(x)



[1, 3, 5, 7, 9]
1
3
5
7
9


In [58]:
# we can create our own generators
# you do this through creating a function that YIELDS values instead of RETURNING values
# the return keyword ends the function
# the yield keyword gives the value yielded but then keeps running
# the yield keyword is your syntax clue that you're dealing with a generator

# lets create our own version of the range function:
def my_range(stop, start=0, step=1):
    while start < stop:
        yield start # <---- tells you this is a generator
        start += step
        
generatedthing = my_range(100, 5, 5)
print(generatedthing)
for x in generatedthing:
    print(x)

<generator object my_range at 0x07AE7568>
5
10
15
20
25
30
35
40
45
50
55
60
65
70
75
80
85
90
95


In [14]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

my_range(20,start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x000002EB1C6E5C48>

#### Infinite Generator

In [59]:
# bad, never create infinite loops

def my_range(stop, start=0, step=1):
    while start:
        yield start # <---- tells you this is a generator
        start += step
        
generatedthing = my_range(100, 5, 5)
print(generatedthing)
for x in generatedthing:
    print(x)

<generator object my_range at 0x07AE7920>
5
10
15
20
25
30
35
40
45
50
55
60
65
70
75
80
85
90
95
100
105
110
115
120
125
130
135
140
145
150
155
160
165
170
175
180
185
190
195
200
205
210
215
220
225
230
235
240
245
250
255
260
265
270
275
280
285
290
295
300
305
310
315
320
325
330
335
340
345
350
355
360
365
370
375
380
385
390
395
400
405
410
415
420
425
430
435
440
445
450
455
460
465
470
475
480
485
490
495
500
505
510
515
520
525
530
535
540
545
550
555
560
565
570
575
580
585
590
595
600
605
610
615
620
625
630
635
640
645
650
655
660
665
670
675
680
685
690
695
700
705
710
715
720
725
730
735
740
745
750
755
760
765
770
775
780
785
790
795
800
805
810
815
820
825
830
835
840
845
850
855
860
865
870
875
880
885
890
895
900
905
910
915
920
925
930
935
940
945
950
955
960
965
970
975
980
985
990
995
1000
1005
1010
1015
1020
1025
1030
1035
1040
1045
1050
1055
1060
1065
1070
1075
1080
1085
1090
1095
1100
1105
1110
1115
1120
1125
1130
1135
1140
1145
1150
1155
1160
1165
1170
1175
11

9685
9690
9695
9700
9705
9710
9715
9720
9725
9730
9735
9740
9745
9750
9755
9760
9765
9770
9775
9780
9785
9790
9795
9800
9805
9810
9815
9820
9825
9830
9835
9840
9845
9850
9855
9860
9865
9870
9875
9880
9885
9890
9895
9900
9905
9910
9915
9920
9925
9930
9935
9940
9945
9950
9955
9960
9965
9970
9975
9980
9985
9990
9995
10000
10005
10010
10015
10020
10025
10030
10035
10040
10045
10050
10055
10060
10065
10070
10075
10080
10085
10090
10095
10100
10105
10110
10115
10120
10125
10130
10135
10140
10145
10150
10155
10160
10165
10170
10175
10180
10185
10190
10195
10200
10205
10210
10215
10220
10225
10230
10235
10240
10245
10250
10255
10260
10265
10270
10275
10280
10285
10290
10295
10300
10305
10310
10315
10320
10325
10330
10335
10340
10345
10350
10355
10360
10365
10370
10375
10380
10385
10390
10395
10400
10405
10410
10415
10420
10425
10430
10435
10440
10445
10450
10455
10460
10465
10470
10475
10480
10485
10490
10495
10500
10505
10510
10515
10520
10525
10530
10535
10540
10545
10550
10555
10560
10565
1

19680
19685
19690
19695
19700
19705
19710
19715
19720
19725
19730
19735
19740
19745
19750
19755
19760
19765
19770
19775
19780
19785
19790
19795
19800
19805
19810
19815
19820
19825
19830
19835
19840
19845
19850
19855
19860
19865
19870
19875
19880
19885
19890
19895
19900
19905
19910
19915
19920
19925
19930
19935
19940
19945
19950
19955
19960
19965
19970
19975
19980
19985
19990
19995
20000
20005
20010
20015
20020
20025
20030
20035
20040
20045
20050
20055
20060
20065
20070
20075
20080
20085
20090
20095
20100
20105
20110
20115
20120
20125
20130
20135
20140
20145
20150
20155
20160
20165
20170
20175
20180
20185
20190
20195
20200
20205
20210
20215
20220
20225
20230
20235
20240
20245
20250
20255
20260
20265
20270
20275
20280
20285
20290
20295
20300
20305
20310
20315
20320
20325
20330
20335
20340
20345
20350
20355
20360
20365
20370
20375
20380
20385
20390
20395
20400
20405
20410
20415
20420
20425
20430
20435
20440
20445
20450
20455
20460
20465
20470
20475
20480
20485
20490
20495
20500
20505
2051

27175
27180
27185
27190
27195
27200
27205
27210
27215
27220
27225
27230
27235
27240
27245
27250
27255
27260
27265
27270
27275
27280
27285
27290
27295
27300
27305
27310
27315
27320
27325
27330
27335
27340
27345
27350
27355
27360
27365
27370
27375
27380
27385
27390
27395
27400
27405
27410
27415
27420
27425
27430
27435
27440
27445
27450
27455
27460
27465
27470
27475
27480
27485
27490
27495
27500
27505
27510
27515
27520
27525
27530
27535
27540
27545
27550
27555
27560
27565
27570
27575
27580
27585
27590
27595
27600
27605
27610
27615
27620
27625
27630
27635
27640
27645
27650
27655
27660
27665
27670
27675
27680
27685
27690
27695
27700
27705
27710
27715
27720
27725
27730
27735
27740
27745
27750
27755
27760
27765
27770
27775
27780
27785
27790
27795
27800
27805
27810
27815
27820
27825
27830
27835
27840
27845
27850
27855
27860
27865
27870
27875
27880
27885
27890
27895
27900
27905
27910
27915
27920
27925
27930
27935
27940
27945
27950
27955
27960
27965
27970
27975
27980
27985
27990
27995
28000
2800

37170
37175
37180
37185
37190
37195
37200
37205
37210
37215
37220
37225
37230
37235
37240
37245
37250
37255
37260
37265
37270
37275
37280
37285
37290
37295
37300
37305
37310
37315
37320
37325
37330
37335
37340
37345
37350
37355
37360
37365
37370
37375
37380
37385
37390
37395
37400
37405
37410
37415
37420
37425
37430
37435
37440
37445
37450
37455
37460
37465
37470
37475
37480
37485
37490
37495
37500
37505
37510
37515
37520
37525
37530
37535
37540
37545
37550
37555
37560
37565
37570
37575
37580
37585
37590
37595
37600
37605
37610
37615
37620
37625
37630
37635
37640
37645
37650
37655
37660
37665
37670
37675
37680
37685
37690
37695
37700
37705
37710
37715
37720
37725
37730
37735
37740
37745
37750
37755
37760
37765
37770
37775
37780
37785
37790
37795
37800
37805
37810
37815
37820
37825
37830
37835
37840
37845
37850
37855
37860
37865
37870
37875
37880
37885
37890
37895
37900
37905
37910
37915
37920
37925
37930
37935
37940
37945
37950
37955
37960
37965
37970
37975
37980
37985
37990
37995
3800

44665
44670
44675
44680
44685
44690
44695
44700
44705
44710
44715
44720
44725
44730
44735
44740
44745
44750
44755
44760
44765
44770
44775
44780
44785
44790
44795
44800
44805
44810
44815
44820
44825
44830
44835
44840
44845
44850
44855
44860
44865
44870
44875
44880
44885
44890
44895
44900
44905
44910
44915
44920
44925
44930
44935
44940
44945
44950
44955
44960
44965
44970
44975
44980
44985
44990
44995
45000
45005
45010
45015
45020
45025
45030
45035
45040
45045
45050
45055
45060
45065
45070
45075
45080
45085
45090
45095
45100
45105
45110
45115
45120
45125
45130
45135
45140
45145
45150
45155
45160
45165
45170
45175
45180
45185
45190
45195
45200
45205
45210
45215
45220
45225
45230
45235
45240
45245
45250
45255
45260
45265
45270
45275
45280
45285
45290
45295
45300
45305
45310
45315
45320
45325
45330
45335
45340
45345
45350
45355
45360
45365
45370
45375
45380
45385
45390
45395
45400
45405
45410
45415
45420
45425
45430
45435
45440
45445
45450
45455
45460
45465
45470
45475
45480
45485
45490
4549

54265
54270
54275
54280
54285
54290
54295
54300
54305
54310
54315
54320
54325
54330
54335
54340
54345
54350
54355
54360
54365
54370
54375
54380
54385
54390
54395
54400
54405
54410
54415
54420
54425
54430
54435
54440
54445
54450
54455
54460
54465
54470
54475
54480
54485
54490
54495
54500
54505
54510
54515
54520
54525
54530
54535
54540
54545
54550
54555
54560
54565
54570
54575
54580
54585
54590
54595
54600
54605
54610
54615
54620
54625
54630
54635
54640
54645
54650
54655
54660
54665
54670
54675
54680
54685
54690
54695
54700
54705
54710
54715
54720
54725
54730
54735
54740
54745
54750
54755
54760
54765
54770
54775
54780
54785
54790
54795
54800
54805
54810
54815
54820
54825
54830
54835
54840
54845
54850
54855
54860
54865
54870
54875
54880
54885
54890
54895
54900
54905
54910
54915
54920
54925
54930
54935
54940
54945
54950
54955
54960
54965
54970
54975
54980
54985
54990
54995
55000
55005
55010
55015
55020
55025
55030
55035
55040
55045
55050
55055
55060
55065
55070
55075
55080
55085
55090
5509

61580
61585
61590
61595
61600
61605
61610
61615
61620
61625
61630
61635
61640
61645
61650
61655
61660
61665
61670
61675
61680
61685
61690
61695
61700
61705
61710
61715
61720
61725
61730
61735
61740
61745
61750
61755
61760
61765
61770
61775
61780
61785
61790
61795
61800
61805
61810
61815
61820
61825
61830
61835
61840
61845
61850
61855
61860
61865
61870
61875
61880
61885
61890
61895
61900
61905
61910
61915
61920
61925
61930
61935
61940
61945
61950
61955
61960
61965
61970
61975
61980
61985
61990
61995
62000
62005
62010
62015
62020
62025
62030
62035
62040
62045
62050
62055
62060
62065
62070
62075
62080
62085
62090
62095
62100
62105
62110
62115
62120
62125
62130
62135
62140
62145
62150
62155
62160
62165
62170
62175
62180
62185
62190
62195
62200
62205
62210
62215
62220
62225
62230
62235
62240
62245
62250
62255
62260
62265
62270
62275
62280
62285
62290
62295
62300
62305
62310
62315
62320
62325
62330
62335
62340
62345
62350
62355
62360
62365
62370
62375
62380
62385
62390
62395
62400
62405
6241

69100
69105
69110
69115
69120
69125
69130
69135
69140
69145
69150
69155
69160
69165
69170
69175
69180
69185
69190
69195
69200
69205
69210
69215
69220
69225
69230
69235
69240
69245
69250
69255
69260
69265
69270
69275
69280
69285
69290
69295
69300
69305
69310
69315
69320
69325
69330
69335
69340
69345
69350
69355
69360
69365
69370
69375
69380
69385
69390
69395
69400
69405
69410
69415
69420
69425
69430
69435
69440
69445
69450
69455
69460
69465
69470
69475
69480
69485
69490
69495
69500
69505
69510
69515
69520
69525
69530
69535
69540
69545
69550
69555
69560
69565
69570
69575
69580
69585
69590
69595
69600
69605
69610
69615
69620
69625
69630
69635
69640
69645
69650
69655
69660
69665
69670
69675
69680
69685
69690
69695
69700
69705
69710
69715
69720
69725
69730
69735
69740
69745
69750
69755
69760
69765
69770
69775
69780
69785
69790
69795
69800
69805
69810
69815
69820
69825
69830
69835
69840
69845
69850
69855
69860
69865
69870
69875
69880
69885
69890
69895
69900
69905
69910
69915
69920
69925
6993

76280
76285
76290
76295
76300
76305
76310
76315
76320
76325
76330
76335
76340
76345
76350
76355
76360
76365
76370
76375
76380
76385
76390
76395
76400
76405
76410
76415
76420
76425
76430
76435
76440
76445
76450
76455
76460
76465
76470
76475
76480
76485
76490
76495
76500
76505
76510
76515
76520
76525
76530
76535
76540
76545
76550
76555
76560
76565
76570
76575
76580
76585
76590
76595
76600
76605
76610
76615
76620
76625
76630
76635
76640
76645
76650
76655
76660
76665
76670
76675
76680
76685
76690
76695
76700
76705
76710
76715
76720
76725
76730
76735
76740
76745
76750
76755
76760
76765
76770
76775
76780
76785
76790
76795
76800
76805
76810
76815
76820
76825
76830
76835
76840
76845
76850
76855
76860
76865
76870
76875
76880
76885
76890
76895
76900
76905
76910
76915
76920
76925
76930
76935
76940
76945
76950
76955
76960
76965
76970
76975
76980
76985
76990
76995
77000
77005
77010
77015
77020
77025
77030
77035
77040
77045
77050
77055
77060
77065
77070
77075
77080
77085
77090
77095
77100
77105
7711

85660
85665
85670
85675
85680
85685
85690
85695
85700
85705
85710
85715
85720
85725
85730
85735
85740
85745
85750
85755
85760
85765
85770
85775
85780
85785
85790
85795
85800
85805
85810
85815
85820
85825
85830
85835
85840
85845
85850
85855
85860
85865
85870
85875
85880
85885
85890
85895
85900
85905
85910
85915
85920
85925
85930
85935
85940
85945
85950
85955
85960
85965
85970
85975
85980
85985
85990
85995
86000
86005
86010
86015
86020
86025
86030
86035
86040
86045
86050
86055
86060
86065
86070
86075
86080
86085
86090
86095
86100
86105
86110
86115
86120
86125
86130
86135
86140
86145
86150
86155
86160
86165
86170
86175
86180
86185
86190
86195
86200
86205
86210
86215
86220
86225
86230
86235
86240
86245
86250
86255
86260
86265
86270
86275
86280
86285
86290
86295
86300
86305
86310
86315
86320
86325
86330
86335
86340
86345
86350
86355
86360
86365
86370
86375
86380
86385
86390
86395
86400
86405
86410
86415
86420
86425
86430
86435
86440
86445
86450
86455
86460
86465
86470
86475
86480
86485
8649

94095
94100
94105
94110
94115
94120
94125
94130
94135
94140
94145
94150
94155
94160
94165
94170
94175
94180
94185
94190
94195
94200
94205
94210
94215
94220
94225
94230
94235
94240
94245
94250
94255
94260
94265
94270
94275
94280
94285
94290
94295
94300
94305
94310
94315
94320
94325
94330
94335
94340
94345
94350
94355
94360
94365
94370
94375
94380
94385
94390
94395
94400
94405
94410
94415
94420
94425
94430
94435
94440
94445
94450
94455
94460
94465
94470
94475
94480
94485
94490
94495
94500
94505
94510
94515
94520
94525
94530
94535
94540
94545
94550
94555
94560
94565
94570
94575
94580
94585
94590
94595
94600
94605
94610
94615
94620
94625
94630
94635
94640
94645
94650
94655
94660
94665
94670
94675
94680
94685
94690
94695
94700
94705
94710
94715
94720
94725
94730
94735
94740
94745
94750
94755
94760
94765
94770
94775
94780
94785
94790
94795
94800
94805
94810
94815
94820
94825
94830
94835
94840
94845
94850
94855
94860
94865
94870
94875
94880
94885
94890
94895
94900
94905
94910
94915
94920
9492

101505
101510
101515
101520
101525
101530
101535
101540
101545
101550
101555
101560
101565
101570
101575
101580
101585
101590
101595
101600
101605
101610
101615
101620
101625
101630
101635
101640
101645
101650
101655
101660
101665
101670
101675
101680
101685
101690
101695
101700
101705
101710
101715
101720
101725
101730
101735
101740
101745
101750
101755
101760
101765
101770
101775
101780
101785
101790
101795
101800
101805
101810
101815
101820
101825
101830
101835
101840
101845
101850
101855
101860
101865
101870
101875
101880
101885
101890
101895
101900
101905
101910
101915
101920
101925
101930
101935
101940
101945
101950
101955
101960
101965
101970
101975
101980
101985
101990
101995
102000
102005
102010
102015
102020
102025
102030
102035
102040
102045
102050
102055
102060
102065
102070
102075
102080
102085
102090
102095
102100
102105
102110
102115
102120
102125
102130
102135
102140
102145
102150
102155
102160
102165
102170
102175
102180
102185
102190
102195
102200
102205
102210
102215

108360
108365
108370
108375
108380
108385
108390
108395
108400
108405
108410
108415
108420
108425
108430
108435
108440
108445
108450
108455
108460
108465
108470
108475
108480
108485
108490
108495
108500
108505
108510
108515
108520
108525
108530
108535
108540
108545
108550
108555
108560
108565
108570
108575
108580
108585
108590
108595
108600
108605
108610
108615
108620
108625
108630
108635
108640
108645
108650
108655
108660
108665
108670
108675
108680
108685
108690
108695
108700
108705
108710
108715
108720
108725
108730
108735
108740
108745
108750
108755
108760
108765
108770
108775
108780
108785
108790
108795
108800
108805
108810
108815
108820
108825
108830
108835
108840
108845
108850
108855
108860
108865
108870
108875
108880
108885
108890
108895
108900
108905
108910
108915
108920
108925
108930
108935
108940
108945
108950
108955
108960
108965
108970
108975
108980
108985
108990
108995
109000
109005
109010
109015
109020
109025
109030
109035
109040
109045
109050
109055
109060
109065
109070

114930
114935
114940
114945
114950
114955
114960
114965
114970
114975
114980
114985
114990
114995
115000
115005
115010
115015
115020
115025
115030
115035
115040
115045
115050
115055
115060
115065
115070
115075
115080
115085
115090
115095
115100
115105
115110
115115
115120
115125
115130
115135
115140
115145
115150
115155
115160
115165
115170
115175
115180
115185
115190
115195
115200
115205
115210
115215
115220
115225
115230
115235
115240
115245
115250
115255
115260
115265
115270
115275
115280
115285
115290
115295
115300
115305
115310
115315
115320
115325
115330
115335
115340
115345
115350
115355
115360
115365
115370
115375
115380
115385
115390
115395
115400
115405
115410
115415
115420
115425
115430
115435
115440
115445
115450
115455
115460
115465
115470
115475
115480
115485
115490
115495
115500
115505
115510
115515
115520
115525
115530
115535
115540
115545
115550
115555
115560
115565
115570
115575
115580
115585
115590
115595
115600
115605
115610
115615
115620
115625
115630
115635
115640

120910
120915
120920
120925
120930
120935
120940
120945
120950
120955
120960
120965
120970
120975
120980
120985
120990
120995
121000
121005
121010
121015
121020
121025
121030
121035
121040
121045
121050
121055
121060
121065
121070
121075
121080
121085
121090
121095
121100
121105
121110
121115
121120
121125
121130
121135
121140
121145
121150
121155
121160
121165
121170
121175
121180
121185
121190
121195
121200
121205
121210
121215
121220
121225
121230
121235
121240
121245
121250
121255
121260
121265
121270
121275
121280
121285
121290
121295
121300
121305
121310
121315
121320
121325
121330
121335
121340
121345
121350
121355
121360
121365
121370
121375
121380
121385
121390
121395
121400
121405
121410
121415
121420
121425
121430
121435
121440
121445
121450
121455
121460
121465
121470
121475
121480
121485
121490
121495
121500
121505
121510
121515
121520
121525
121530
121535
121540
121545
121550
121555
121560
121565
121570
121575
121580
121585
121590
121595
121600
121605
121610
121615
121620

129610
129615
129620
129625
129630
129635
129640
129645
129650
129655
129660
129665
129670
129675
129680
129685
129690
129695
129700
129705
129710
129715
129720
129725
129730
129735
129740
129745
129750
129755
129760
129765
129770
129775
129780
129785
129790
129795
129800
129805
129810
129815
129820
129825
129830
129835
129840
129845
129850
129855
129860
129865
129870
129875
129880
129885
129890
129895
129900
129905
129910
129915
129920
129925
129930
129935
129940
129945
129950
129955
129960
129965
129970
129975
129980
129985
129990
129995
130000
130005
130010
130015
130020
130025
130030
130035
130040
130045
130050
130055
130060
130065
130070
130075
130080
130085
130090
130095
130100
130105
130110
130115
130120
130125
130130
130135
130140
130145
130150
130155
130160
130165
130170
130175
130180
130185
130190
130195
130200
130205
130210
130215
130220
130225
130230
130235
130240
130245
130250
130255
130260
130265
130270
130275
130280
130285
130290
130295
130300
130305
130310
130315
130320

139605
139610
139615
139620
139625
139630
139635
139640
139645
139650
139655
139660
139665
139670
139675
139680
139685
139690
139695
139700
139705
139710
139715
139720
139725
139730
139735
139740
139745
139750
139755
139760
139765
139770
139775
139780
139785
139790
139795
139800
139805
139810
139815
139820
139825
139830
139835
139840
139845
139850
139855
139860
139865
139870
139875
139880
139885
139890
139895
139900
139905
139910
139915
139920
139925
139930
139935
139940
139945
139950
139955
139960
139965
139970
139975
139980
139985
139990
139995
140000
140005
140010
140015
140020
140025
140030
140035
140040
140045
140050
140055
140060
140065
140070
140075
140080
140085
140090
140095
140100
140105
140110
140115
140120
140125
140130
140135
140140
140145
140150
140155
140160
140165
140170
140175
140180
140185
140190
140195
140200
140205
140210
140215
140220
140225
140230
140235
140240
140245
140250
140255
140260
140265
140270
140275
140280
140285
140290
140295
140300
140305
140310
140315

147100
147105
147110
147115
147120
147125
147130
147135
147140
147145
147150
147155
147160
147165
147170
147175
147180
147185
147190
147195
147200
147205
147210
147215
147220
147225
147230
147235
147240
147245
147250
147255
147260
147265
147270
147275
147280
147285
147290
147295
147300
147305
147310
147315
147320
147325
147330
147335
147340
147345
147350
147355
147360
147365
147370
147375
147380
147385
147390
147395
147400
147405
147410
147415
147420
147425
147430
147435
147440
147445
147450
147455
147460
147465
147470
147475
147480
147485
147490
147495
147500
147505
147510
147515
147520
147525
147530
147535
147540
147545
147550
147555
147560
147565
147570
147575
147580
147585
147590
147595
147600
147605
147610
147615
147620
147625
147630
147635
147640
147645
147650
147655
147660
147665
147670
147675
147680
147685
147690
147695
147700
147705
147710
147715
147720
147725
147730
147735
147740
147745
147750
147755
147760
147765
147770
147775
147780
147785
147790
147795
147800
147805
147810

154595
154600
154605
154610
154615
154620
154625
154630
154635
154640
154645
154650
154655
154660
154665
154670
154675
154680
154685
154690
154695
154700
154705
154710
154715
154720
154725
154730
154735
154740
154745
154750
154755
154760
154765
154770
154775
154780
154785
154790
154795
154800
154805
154810
154815
154820
154825
154830
154835
154840
154845
154850
154855
154860
154865
154870
154875
154880
154885
154890
154895
154900
154905
154910
154915
154920
154925
154930
154935
154940
154945
154950
154955
154960
154965
154970
154975
154980
154985
154990
154995
155000
155005
155010
155015
155020
155025
155030
155035
155040
155045
155050
155055
155060
155065
155070
155075
155080
155085
155090
155095
155100
155105
155110
155115
155120
155125
155130
155135
155140
155145
155150
155155
155160
155165
155170
155175
155180
155185
155190
155195
155200
155205
155210
155215
155220
155225
155230
155235
155240
155245
155250
155255
155260
155265
155270
155275
155280
155285
155290
155295
155300
155305

162090
162095
162100
162105
162110
162115
162120
162125
162130
162135
162140
162145
162150
162155
162160
162165
162170
162175
162180
162185
162190
162195
162200
162205
162210
162215
162220
162225
162230
162235
162240
162245
162250
162255
162260
162265
162270
162275
162280
162285
162290
162295
162300
162305
162310
162315
162320
162325
162330
162335
162340
162345
162350
162355
162360
162365
162370
162375
162380
162385
162390
162395
162400
162405
162410
162415
162420
162425
162430
162435
162440
162445
162450
162455
162460
162465
162470
162475
162480
162485
162490
162495
162500
162505
162510
162515
162520
162525
162530
162535
162540
162545
162550
162555
162560
162565
162570
162575
162580
162585
162590
162595
162600
162605
162610
162615
162620
162625
162630
162635
162640
162645
162650
162655
162660
162665
162670
162675
162680
162685
162690
162695
162700
162705
162710
162715
162720
162725
162730
162735
162740
162745
162750
162755
162760
162765
162770
162775
162780
162785
162790
162795
162800

169585
169590
169595
169600
169605
169610
169615
169620
169625
169630
169635
169640
169645
169650
169655
169660
169665
169670
169675
169680
169685
169690
169695
169700
169705
169710
169715
169720
169725
169730
169735
169740
169745
169750
169755
169760
169765
169770
169775
169780
169785
169790
169795
169800
169805
169810
169815
169820
169825
169830
169835
169840
169845
169850
169855
169860
169865
169870
169875
169880
169885
169890
169895
169900
169905
169910
169915
169920
169925
169930
169935
169940
169945
169950
169955
169960
169965
169970
169975
169980
169985
169990
169995
170000
170005
170010
170015
170020
170025
170030
170035
170040
170045
170050
170055
170060
170065
170070
170075
170080
170085
170090
170095
170100
170105
170110
170115
170120
170125
170130
170135
170140
170145
170150
170155
170160
170165
170170
170175
170180
170185
170190
170195
170200
170205
170210
170215
170220
170225
170230
170235
170240
170245
170250
170255
170260
170265
170270
170275
170280
170285
170290
170295

177080
177085
177090
177095
177100
177105
177110
177115
177120
177125
177130
177135
177140
177145
177150
177155
177160
177165
177170
177175
177180
177185
177190
177195
177200
177205
177210
177215
177220
177225
177230
177235
177240
177245
177250
177255
177260
177265
177270
177275
177280
177285
177290
177295
177300
177305
177310
177315
177320
177325
177330
177335
177340
177345
177350
177355
177360
177365
177370
177375
177380
177385
177390
177395
177400
177405
177410
177415
177420
177425
177430
177435
177440
177445
177450
177455
177460
177465
177470
177475
177480
177485
177490
177495
177500
177505
177510
177515
177520
177525
177530
177535
177540
177545
177550
177555
177560
177565
177570
177575
177580
177585
177590
177595
177600
177605
177610
177615
177620
177625
177630
177635
177640
177645
177650
177655
177660
177665
177670
177675
177680
177685
177690
177695
177700
177705
177710
177715
177720
177725
177730
177735
177740
177745
177750
177755
177760
177765
177770
177775
177780
177785
177790

184575
184580
184585
184590
184595
184600
184605
184610
184615
184620
184625
184630
184635
184640
184645
184650
184655
184660
184665
184670
184675
184680
184685
184690
184695
184700
184705
184710
184715
184720
184725
184730
184735
184740
184745
184750
184755
184760
184765
184770
184775
184780
184785
184790
184795
184800
184805
184810
184815
184820
184825
184830
184835
184840
184845
184850
184855
184860
184865
184870
184875
184880
184885
184890
184895
184900
184905
184910
184915
184920
184925
184930
184935
184940
184945
184950
184955
184960
184965
184970
184975
184980
184985
184990
184995
185000
185005
185010
185015
185020
185025
185030
185035
185040
185045
185050
185055
185060
185065
185070
185075
185080
185085
185090
185095
185100
185105
185110
185115
185120
185125
185130
185135
185140
185145
185150
185155
185160
185165
185170
185175
185180
185185
185190
185195
185200
185205
185210
185215
185220
185225
185230
185235
185240
185245
185250
185255
185260
185265
185270
185275
185280
185285

191750
191755
191760
191765
191770
191775
191780
191785
191790
191795
191800
191805
191810
191815
191820
191825
191830
191835
191840
191845
191850
191855
191860
191865
191870
191875
191880
191885
191890
191895
191900
191905
191910
191915
191920
191925
191930
191935
191940
191945
191950
191955
191960
191965
191970
191975
191980
191985
191990
191995
192000
192005
192010
192015
192020
192025
192030
192035
192040
192045
192050
192055
192060
192065
192070
192075
192080
192085
192090
192095
192100
192105
192110
192115
192120
192125
192130
192135
192140
192145
192150
192155
192160
192165
192170
192175
192180
192185
192190
192195
192200
192205
192210
192215
192220
192225
192230
192235
192240
192245
192250
192255
192260
192265
192270
192275
192280
192285
192290
192295
192300
192305
192310
192315
192320
192325
192330
192335
192340
192345
192350
192355
192360
192365
192370
192375
192380
192385
192390
192395
192400
192405
192410
192415
192420
192425
192430
192435
192440
192445
192450
192455
192460

199565
199570
199575
199580
199585
199590
199595
199600
199605
199610
199615
199620
199625
199630
199635
199640
199645
199650
199655
199660
199665
199670
199675
199680
199685
199690
199695
199700
199705
199710
199715
199720
199725
199730
199735
199740
199745
199750
199755
199760
199765
199770
199775
199780
199785
199790
199795
199800
199805
199810
199815
199820
199825
199830
199835
199840
199845
199850
199855
199860
199865
199870
199875
199880
199885
199890
199895
199900
199905
199910
199915
199920
199925
199930
199935
199940
199945
199950
199955
199960
199965
199970
199975
199980
199985
199990
199995
200000
200005
200010
200015
200020
200025
200030
200035
200040
200045
200050
200055
200060
200065
200070
200075
200080
200085
200090
200095
200100
200105
200110
200115
200120
200125
200130
200135
200140
200145
200150
200155
200160
200165
200170
200175
200180
200185
200190
200195
200200
200205
200210
200215
200220
200225
200230
200235
200240
200245
200250
200255
200260
200265
200270
200275

207060
207065
207070
207075
207080
207085
207090
207095
207100
207105
207110
207115
207120
207125
207130
207135
207140
207145
207150
207155
207160
207165
207170
207175
207180
207185
207190
207195
207200
207205
207210
207215
207220
207225
207230
207235
207240
207245
207250
207255
207260
207265
207270
207275
207280
207285
207290
207295
207300
207305
207310
207315
207320
207325
207330
207335
207340
207345
207350
207355
207360
207365
207370
207375
207380
207385
207390
207395
207400
207405
207410
207415
207420
207425
207430
207435
207440
207445
207450
207455
207460
207465
207470
207475
207480
207485
207490
207495
207500
207505
207510
207515
207520
207525
207530
207535
207540
207545
207550
207555
207560
207565
207570
207575
207580
207585
207590
207595
207600
207605
207610
207615
207620
207625
207630
207635
207640
207645
207650
207655
207660
207665
207670
207675
207680
207685
207690
207695
207700
207705
207710
207715
207720
207725
207730
207735
207740
207745
207750
207755
207760
207765
207770

214870
214875
214880
214885
214890
214895
214900
214905
214910
214915
214920
214925
214930
214935
214940
214945
214950
214955
214960
214965
214970
214975
214980
214985
214990
214995
215000
215005
215010
215015
215020
215025
215030
215035
215040
215045
215050
215055
215060
215065
215070
215075
215080
215085
215090
215095
215100
215105
215110
215115
215120
215125
215130
215135
215140
215145
215150
215155
215160
215165
215170
215175
215180
215185
215190
215195
215200
215205
215210
215215
215220
215225
215230
215235
215240
215245
215250
215255
215260
215265
215270
215275
215280
215285
215290
215295
215300
215305
215310
215315
215320
215325
215330
215335
215340
215345
215350
215355
215360
215365
215370
215375
215380
215385
215390
215395
215400
215405
215410
215415
215420
215425
215430
215435
215440
215445
215450
215455
215460
215465
215470
215475
215480
215485
215490
215495
215500
215505
215510
215515
215520
215525
215530
215535
215540
215545
215550
215555
215560
215565
215570
215575
215580

222010
222015
222020
222025
222030
222035
222040
222045
222050
222055
222060
222065
222070
222075
222080
222085
222090
222095
222100
222105
222110
222115
222120
222125
222130
222135
222140
222145
222150
222155
222160
222165
222170
222175
222180
222185
222190
222195
222200
222205
222210
222215
222220
222225
222230
222235
222240
222245
222250
222255
222260
222265
222270
222275
222280
222285
222290
222295
222300
222305
222310
222315
222320
222325
222330
222335
222340
222345
222350
222355
222360
222365
222370
222375
222380
222385
222390
222395
222400
222405
222410
222415
222420
222425
222430
222435
222440
222445
222450
222455
222460
222465
222470
222475
222480
222485
222490
222495
222500
222505
222510
222515
222520
222525
222530
222535
222540
222545
222550
222555
222560
222565
222570
222575
222580
222585
222590
222595
222600
222605
222610
222615
222620
222625
222630
222635
222640
222645
222650
222655
222660
222665
222670
222675
222680
222685
222690
222695
222700
222705
222710
222715
222720

229545
229550
229555
229560
229565
229570
229575
229580
229585
229590
229595
229600
229605
229610
229615
229620
229625
229630
229635
229640
229645
229650
229655
229660
229665
229670
229675
229680
229685
229690
229695
229700
229705
229710
229715
229720
229725
229730
229735
229740
229745
229750
229755
229760
229765
229770
229775
229780
229785
229790
229795
229800
229805
229810
229815
229820
229825
229830
229835
229840
229845
229850
229855
229860
229865
229870
229875
229880
229885
229890
229895
229900
229905
229910
229915
229920
229925
229930
229935
229940
229945
229950
229955
229960
229965
229970
229975
229980
229985
229990
229995
230000
230005
230010
230015
230020
230025
230030
230035
230040
230045
230050
230055
230060
230065
230070
230075
230080
230085
230090
230095
230100
230105
230110
230115
230120
230125
230130
230135
230140
230145
230150
230155
230160
230165
230170
230175
230180
230185
230190
230195
230200
230205
230210
230215
230220
230225
230230
230235
230240
230245
230250
230255

236955
236960
236965
236970
236975
236980
236985
236990
236995
237000
237005
237010
237015
237020
237025
237030
237035
237040
237045
237050
237055
237060
237065
237070
237075
237080
237085
237090
237095
237100
237105
237110
237115
237120
237125
237130
237135
237140
237145
237150
237155
237160
237165
237170
237175
237180
237185
237190
237195
237200
237205
237210
237215
237220
237225
237230
237235
237240
237245
237250
237255
237260
237265
237270
237275
237280
237285
237290
237295
237300
237305
237310
237315
237320
237325
237330
237335
237340
237345
237350
237355
237360
237365
237370
237375
237380
237385
237390
237395
237400
237405
237410
237415
237420
237425
237430
237435
237440
237445
237450
237455
237460
237465
237470
237475
237480
237485
237490
237495
237500
237505
237510
237515
237520
237525
237530
237535
237540
237545
237550
237555
237560
237565
237570
237575
237580
237585
237590
237595
237600
237605
237610
237615
237620
237625
237630
237635
237640
237645
237650
237655
237660
237665

245185
245190
245195
245200
245205
245210
245215
245220
245225
245230
245235
245240
245245
245250
245255
245260
245265
245270
245275
245280
245285
245290
245295
245300
245305
245310
245315
245320
245325
245330
245335
245340
245345
245350
245355
245360
245365
245370
245375
245380
245385
245390
245395
245400
245405
245410
245415
245420
245425
245430
245435
245440
245445
245450
245455
245460
245465
245470
245475
245480
245485
245490
245495
245500
245505
245510
245515
245520
245525
245530
245535
245540
245545
245550
245555
245560
245565
245570
245575
245580
245585
245590
245595
245600
245605
245610
245615
245620
245625
245630
245635
245640
245645
245650
245655
245660
245665
245670
245675
245680
245685
245690
245695
245700
245705
245710
245715
245720
245725
245730
245735
245740
245745
245750
245755
245760
245765
245770
245775
245780
245785
245790
245795
245800
245805
245810
245815
245820
245825
245830
245835
245840
245845
245850
245855
245860
245865
245870
245875
245880
245885
245890
245895

251600
251605
251610
251615
251620
251625
251630
251635
251640
251645
251650
251655
251660
251665
251670
251675
251680
251685
251690
251695
251700
251705
251710
251715
251720
251725
251730
251735
251740
251745
251750
251755
251760
251765
251770
251775
251780
251785
251790
251795
251800
251805
251810
251815
251820
251825
251830
251835
251840
251845
251850
251855
251860
251865
251870
251875
251880
251885
251890
251895
251900
251905
251910
251915
251920
251925
251930
251935
251940
251945
251950
251955
251960
251965
251970
251975
251980
251985
251990
251995
252000
252005
252010
252015
252020
252025
252030
252035
252040
252045
252050
252055
252060
252065
252070
252075
252080
252085
252090
252095
252100
252105
252110
252115
252120
252125
252130
252135
252140
252145
252150
252155
252160
252165
252170
252175
252180
252185
252190
252195
252200
252205
252210
252215
252220
252225
252230
252235
252240
252245
252250
252255
252260
252265
252270
252275
252280
252285
252290
252295
252300
252305
252310

259520
259525
259530
259535
259540
259545
259550
259555
259560
259565
259570
259575
259580
259585
259590
259595
259600
259605
259610
259615
259620
259625
259630
259635
259640
259645
259650
259655
259660
259665
259670
259675
259680
259685
259690
259695
259700
259705
259710
259715
259720
259725
259730
259735
259740
259745
259750
259755
259760
259765
259770
259775
259780
259785
259790
259795
259800
259805
259810
259815
259820
259825
259830
259835
259840
259845
259850
259855
259860
259865
259870
259875
259880
259885
259890
259895
259900
259905
259910
259915
259920
259925
259930
259935
259940
259945
259950
259955
259960
259965
259970
259975
259980
259985
259990
259995
260000
260005
260010
260015
260020
260025
260030
260035
260040
260045
260050
260055
260060
260065
260070
260075
260080
260085
260090
260095
260100
260105
260110
260115
260120
260125
260130
260135
260140
260145
260150
260155
260160
260165
260170
260175
260180
260185
260190
260195
260200
260205
260210
260215
260220
260225
260230

269515
269520
269525
269530
269535
269540
269545
269550
269555
269560
269565
269570
269575
269580
269585
269590
269595
269600
269605
269610
269615
269620
269625
269630
269635
269640
269645
269650
269655
269660
269665
269670
269675
269680
269685
269690
269695
269700
269705
269710
269715
269720
269725
269730
269735
269740
269745
269750
269755
269760
269765
269770
269775
269780
269785
269790
269795
269800
269805
269810
269815
269820
269825
269830
269835
269840
269845
269850
269855
269860
269865
269870
269875
269880
269885
269890
269895
269900
269905
269910
269915
269920
269925
269930
269935
269940
269945
269950
269955
269960
269965
269970
269975
269980
269985
269990
269995
270000
270005
270010
270015
270020
270025
270030
270035
270040
270045
270050
270055
270060
270065
270070
270075
270080
270085
270090
270095
270100
270105
270110
270115
270120
270125
270130
270135
270140
270145
270150
270155
270160
270165
270170
270175
270180
270185
270190
270195
270200
270205
270210
270215
270220
270225

277010
277015
277020
277025
277030
277035
277040
277045
277050
277055
277060
277065
277070
277075
277080
277085
277090
277095
277100
277105
277110
277115
277120
277125
277130
277135
277140
277145
277150
277155
277160
277165
277170
277175
277180
277185
277190
277195
277200
277205
277210
277215
277220
277225
277230
277235
277240
277245
277250
277255
277260
277265
277270
277275
277280
277285
277290
277295
277300
277305
277310
277315
277320
277325
277330
277335
277340
277345
277350
277355
277360
277365
277370
277375
277380
277385
277390
277395
277400
277405
277410
277415
277420
277425
277430
277435
277440
277445
277450
277455
277460
277465
277470
277475
277480
277485
277490
277495
277500
277505
277510
277515
277520
277525
277530
277535
277540
277545
277550
277555
277560
277565
277570
277575
277580
277585
277590
277595
277600
277605
277610
277615
277620
277625
277630
277635
277640
277645
277650
277655
277660
277665
277670
277675
277680
277685
277690
277695
277700
277705
277710
277715
277720

284505
284510
284515
284520
284525
284530
284535
284540
284545
284550
284555
284560
284565
284570
284575
284580
284585
284590
284595
284600
284605
284610
284615
284620
284625
284630
284635
284640
284645
284650
284655
284660
284665
284670
284675
284680
284685
284690
284695
284700
284705
284710
284715
284720
284725
284730
284735
284740
284745
284750
284755
284760
284765
284770
284775
284780
284785
284790
284795
284800
284805
284810
284815
284820
284825
284830
284835
284840
284845
284850
284855
284860
284865
284870
284875
284880
284885
284890
284895
284900
284905
284910
284915
284920
284925
284930
284935
284940
284945
284950
284955
284960
284965
284970
284975
284980
284985
284990
284995
285000
285005
285010
285015
285020
285025
285030
285035
285040
285045
285050
285055
285060
285065
285070
285075
285080
285085
285090
285095
285100
285105
285110
285115
285120
285125
285130
285135
285140
285145
285150
285155
285160
285165
285170
285175
285180
285185
285190
285195
285200
285205
285210
285215

292000
292005
292010
292015
292020
292025
292030
292035
292040
292045
292050
292055
292060
292065
292070
292075
292080
292085
292090
292095
292100
292105
292110
292115
292120
292125
292130
292135
292140
292145
292150
292155
292160
292165
292170
292175
292180
292185
292190
292195
292200
292205
292210
292215
292220
292225
292230
292235
292240
292245
292250
292255
292260
292265
292270
292275
292280
292285
292290
292295
292300
292305
292310
292315
292320
292325
292330
292335
292340
292345
292350
292355
292360
292365
292370
292375
292380
292385
292390
292395
292400
292405
292410
292415
292420
292425
292430
292435
292440
292445
292450
292455
292460
292465
292470
292475
292480
292485
292490
292495
292500
292505
292510
292515
292520
292525
292530
292535
292540
292545
292550
292555
292560
292565
292570
292575
292580
292585
292590
292595
292600
292605
292610
292615
292620
292625
292630
292635
292640
292645
292650
292655
292660
292665
292670
292675
292680
292685
292690
292695
292700
292705
292710

299235
299240
299245
299250
299255
299260
299265
299270
299275
299280
299285
299290
299295
299300
299305
299310
299315
299320
299325
299330
299335
299340
299345
299350
299355
299360
299365
299370
299375
299380
299385
299390
299395
299400
299405
299410
299415
299420
299425
299430
299435
299440
299445
299450
299455
299460
299465
299470
299475
299480
299485
299490
299495
299500
299505
299510
299515
299520
299525
299530
299535
299540
299545
299550
299555
299560
299565
299570
299575
299580
299585
299590
299595
299600
299605
299610
299615
299620
299625
299630
299635
299640
299645
299650
299655
299660
299665
299670
299675
299680
299685
299690
299695
299700
299705
299710
299715
299720
299725
299730
299735
299740
299745
299750
299755
299760
299765
299770
299775
299780
299785
299790
299795
299800
299805
299810
299815
299820
299825
299830
299835
299840
299845
299850
299855
299860
299865
299870
299875
299880
299885
299890
299895
299900
299905
299910
299915
299920
299925
299930
299935
299940
299945

306045
306050
306055
306060
306065
306070
306075
306080
306085
306090
306095
306100
306105
306110
306115
306120
306125
306130
306135
306140
306145
306150
306155
306160
306165
306170
306175
306180
306185
306190
306195
306200
306205
306210
306215
306220
306225
306230
306235
306240
306245
306250
306255
306260
306265
306270
306275
306280
306285
306290
306295
306300
306305
306310
306315
306320
306325
306330
306335
306340
306345
306350
306355
306360
306365
306370
306375
306380
306385
306390
306395
306400
306405
306410
306415
306420
306425
306430
306435
306440
306445
306450
306455
306460
306465
306470
306475
306480
306485
306490
306495
306500
306505
306510
306515
306520
306525
306530
306535
306540
306545
306550
306555
306560
306565
306570
306575
306580
306585
306590
306595
306600
306605
306610
306615
306620
306625
306630
306635
306640
306645
306650
306655
306660
306665
306670
306675
306680
306685
306690
306695
306700
306705
306710
306715
306720
306725
306730
306735
306740
306745
306750
306755

313695
313700
313705
313710
313715
313720
313725
313730
313735
313740
313745
313750
313755
313760
313765
313770
313775
313780
313785
313790
313795
313800
313805
313810
313815
313820
313825
313830
313835
313840
313845
313850
313855
313860
313865
313870
313875
313880
313885
313890
313895
313900
313905
313910
313915
313920
313925
313930
313935
313940
313945
313950
313955
313960
313965
313970
313975
313980
313985
313990
313995
314000
314005
314010
314015
314020
314025
314030
314035
314040
314045
314050
314055
314060
314065
314070
314075
314080
314085
314090
314095
314100
314105
314110
314115
314120
314125
314130
314135
314140
314145
314150
314155
314160
314165
314170
314175
314180
314185
314190
314195
314200
314205
314210
314215
314220
314225
314230
314235
314240
314245
314250
314255
314260
314265
314270
314275
314280
314285
314290
314295
314300
314305
314310
314315
314320
314325
314330
314335
314340
314345
314350
314355
314360
314365
314370
314375
314380
314385
314390
314395
314400
314405

320340
320345
320350
320355
320360
320365
320370
320375
320380
320385
320390
320395
320400
320405
320410
320415
320420
320425
320430
320435
320440
320445
320450
320455
320460
320465
320470
320475
320480
320485
320490
320495
320500
320505
320510
320515
320520
320525
320530
320535
320540
320545
320550
320555
320560
320565
320570
320575
320580
320585
320590
320595
320600
320605
320610
320615
320620
320625
320630
320635
320640
320645
320650
320655
320660
320665
320670
320675
320680
320685
320690
320695
320700
320705
320710
320715
320720
320725
320730
320735
320740
320745
320750
320755
320760
320765
320770
320775
320780
320785
320790
320795
320800
320805
320810
320815
320820
320825
320830
320835
320840
320845
320850
320855
320860
320865
320870
320875
320880
320885
320890
320895
320900
320905
320910
320915
320920
320925
320930
320935
320940
320945
320950
320955
320960
320965
320970
320975
320980
320985
320990
320995
321000
321005
321010
321015
321020
321025
321030
321035
321040
321045
321050

327380
327385
327390
327395
327400
327405
327410
327415
327420
327425
327430
327435
327440
327445
327450
327455
327460
327465
327470
327475
327480
327485
327490
327495
327500
327505
327510
327515
327520
327525
327530
327535
327540
327545
327550
327555
327560
327565
327570
327575
327580
327585
327590
327595
327600
327605
327610
327615
327620
327625
327630
327635
327640
327645
327650
327655
327660
327665
327670
327675
327680
327685
327690
327695
327700
327705
327710
327715
327720
327725
327730
327735
327740
327745
327750
327755
327760
327765
327770
327775
327780
327785
327790
327795
327800
327805
327810
327815
327820
327825
327830
327835
327840
327845
327850
327855
327860
327865
327870
327875
327880
327885
327890
327895
327900
327905
327910
327915
327920
327925
327930
327935
327940
327945
327950
327955
327960
327965
327970
327975
327980
327985
327990
327995
328000
328005
328010
328015
328020
328025
328030
328035
328040
328045
328050
328055
328060
328065
328070
328075
328080
328085
328090

334155
334160
334165
334170
334175
334180
334185
334190
334195
334200
334205
334210
334215
334220
334225
334230
334235
334240
334245
334250
334255
334260
334265
334270
334275
334280
334285
334290
334295
334300
334305
334310
334315
334320
334325
334330
334335
334340
334345
334350
334355
334360
334365
334370
334375
334380
334385
334390
334395
334400
334405
334410
334415
334420
334425
334430
334435
334440
334445
334450
334455
334460
334465
334470
334475
334480
334485
334490
334495
334500
334505
334510
334515
334520
334525
334530
334535
334540
334545
334550
334555
334560
334565
334570
334575
334580
334585
334590
334595
334600
334605
334610
334615
334620
334625
334630
334635
334640
334645
334650
334655
334660
334665
334670
334675
334680
334685
334690
334695
334700
334705
334710
334715
334720
334725
334730
334735
334740
334745
334750
334755
334760
334765
334770
334775
334780
334785
334790
334795
334800
334805
334810
334815
334820
334825
334830
334835
334840
334845
334850
334855
334860
334865

341240
341245
341250
341255
341260
341265
341270
341275
341280
341285
341290
341295
341300
341305
341310
341315
341320
341325
341330
341335
341340
341345
341350
341355
341360
341365
341370
341375
341380
341385
341390
341395
341400
341405
341410
341415
341420
341425
341430
341435
341440
341445
341450
341455
341460
341465
341470
341475
341480
341485
341490
341495
341500
341505
341510
341515
341520
341525
341530
341535
341540
341545
341550
341555
341560
341565
341570
341575
341580
341585
341590
341595
341600
341605
341610
341615
341620
341625
341630
341635
341640
341645
341650
341655
341660
341665
341670
341675
341680
341685
341690
341695
341700
341705
341710
341715
341720
341725
341730
341735
341740
341745
341750
341755
341760
341765
341770
341775
341780
341785
341790
341795
341800
341805
341810
341815
341820
341825
341830
341835
341840
341845
341850
341855
341860
341865
341870
341875
341880
341885
341890
341895
341900
341905
341910
341915
341920
341925
341930
341935
341940
341945
341950

KeyboardInterrupt: 

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [92]:
# for x in generator(4) -> 16, 9, 4, 1, 0

def generator(num):
    while num >= 0:
        yield num**2 # <---- tells you this is a generator
        num-=1
        
for x in generator(4):
    print(x)


16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [128]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

print(list(filter(str.strip, places)))

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [2]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.split(" ")[-1].lower())
print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [17]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

#F_C = (9/5)*int(c) + 32
f = list(map(lambda x: (x[0], (9/5)*x[-1] + 32), places))

print(f)

         
#x[0], x[-1]

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


In [21]:
new_places = ["chicago", "boston", "new york", "salt lake city", "miami"]
print(new_places[-1])

miami


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [30]:
def fib(n):
    if n <= 1:
        print('Iteration(1) = 1')
        return n
    else:
        print(f'Iteration({n}): {n} + ({n+1})')
        return(n*n+1+n*n)

fib(5)

Iteration(5): 5 + (6)


51